src/test/java/co/rsk/bitcoinj/core/TransactionWitnessTest.java

```diff
@@ -186,11 +202,231 @@ public void equals_withTwoTransactionWitnessesWithOneDifferentPush_shouldBeFalse
         transactionWitness1.setPush(0, samePush);
         transactionWitness2.setPush(0, anotherDifferentPush);
 
+        // act
         int hashCode1 = transactionWitness1.hashCode();
         int hashCode2 = transactionWitness2.hashCode();
 
         // assert
         assertNotEquals(transactionWitness1, transactionWitness2);
         assertNotEquals(hashCode1, hashCode2);
     }
+
+    @Test
+    public void getSigInsertionIndex_whenEmptyWitness_shouldThrownArrayIndexOutOfBoundsException() {
+        // arrange
+        pushes = new ArrayList<>();
+        TransactionWitness transactionWitness = TransactionWitness.of(pushes);
+
+        // act & assert
+        assertThrows(ArrayIndexOutOfBoundsException.class, () -> transactionWitness.getSigInsertionIndex(hashForSignature, fedKey1));
+    }
+
+    @Test
+    public void getSigInsertionIndex_whenMalformedRedeemScript_shouldThrowException() {
+        // arrange
+        Script customRedeemScript = new Script(new byte[2]);
+        byte[] emptyByte = {};
+        pushes = new ArrayList<>();
+        pushes.add(emptyByte);
+        pushes.add(customRedeemScript.getProgram());
+        TransactionWitness transactionWitness = TransactionWitness.of(pushes);
+
+        // act & assert
+        assertThrows(ScriptException.class, () -> transactionWitness.getSigInsertionIndex(hashForSignature, fedKey1));
+    }
+
+    @Test
+    public void getSigInsertionIndex_withWitnessWithoutSignatures_shouldReturnZero() {
+        // arrange
+        BtcTransaction btcTx = getBtcTransactionWithBaseWitnessInInput();
+        Sha256Hash btcTxSigHash = btcTx.hashForWitnessSignature(FIRST_INPUT_INDEX, redeemScript, prevValue,
+            BtcTransaction.SigHash.ALL, false);
+        TransactionWitness transactionWitness = btcTx.getWitness(FIRST_INPUT_INDEX);
+
+        // act
+        for (BtcECKey key: FEDERATION_KEYS) {
+            int sigInsertionIndex = transactionWitness.getSigInsertionIndex(btcTxSigHash, key);
+
+            // assert
+            Assert.assertEquals(0, sigInsertionIndex);
+        }
+    }
+
+    @Test
+    public void getSigInsertionIndex_withAGreaterPubKeySignatureInTheWitness_shouldReturnIndexCorrectly() {
+        // arrange
+        BtcTransaction btcTx = getBtcTransactionWithBaseWitnessInInput();
+        Sha256Hash btcTxSigHash = btcTx.hashForWitnessSignature(FIRST_INPUT_INDEX, redeemScript, prevValue,
+            BtcTransaction.SigHash.ALL, false);
+
+        // sign with fedKey1
+        signInput(btcTx, fedKey1, FIRST_INPUT_INDEX, btcTxSigHash);
```

Comment:
Check sig insertion index for both keys before signing? And check again for both keys after signing

Reply:
In this case, the test `getSigInsertionIndex_withTheSamePubKeyWhichSignatureAlreadyInTheWitness_shouldReturnIndexCorrectly` doesn't make sense, right?